compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1863,23 +1863,26 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     return true;
                 }
             }
-            hir::ExprKind::Struct(qpath, fields, ..) => {
-                if let Res::Def(DefKind::Struct | DefKind::Variant, variant_def_id) =
+            hir::ExprKind::Struct(qpath, _fields, ..) => {
```

Comment:
Another thing you can do after the rest of the review is done -- all of these methods under `adjust_fulfillment_errors_for_expr_obligation` can be moved into a distinct module. Maybe `adjust_fulfillment_errors.rs` or something.

Reply:
I ran into some merge conflicts trying to move the existing functions - the new ones are currently placed into a separate module. I'd like to move the remaining ones in a separate PR, just to avoid rebase-churn on this one, since it adds a lot of net-new code.